sdk/core/core/src/az_json_builder.c

```diff
@@ -2,220 +2,627 @@
-      remaining_json = az_span_copy(remaining_json, AZ_SPAN_FROM_STR("null"));
-      json_builder->_internal.length += 4;
-      break;
-    }
-    case AZ_JSON_TOKEN_BOOLEAN:
-    {
-      az_span boolean_literal_string
-          = token._internal.boolean ? AZ_SPAN_FROM_STR("true") : AZ_SPAN_FROM_STR("false");
-      AZ_RETURN_IF_NOT_ENOUGH_SIZE(remaining_json, az_span_size(boolean_literal_string));
-      json_builder->_internal.need_comma = true;
-      remaining_json = az_span_copy(remaining_json, boolean_literal_string);
-      json_builder->_internal.length += az_span_size(boolean_literal_string);
-      break;
-    }
-    case AZ_JSON_TOKEN_NUMBER:
-    {
-      json_builder->_internal.need_comma = true;
-      az_span remainder;
-      AZ_RETURN_IF_FAILED(az_span_dtoa(remaining_json, token._internal.number, &remainder));
-      json_builder->_internal.length += _az_span_diff(remainder, remaining_json);
-      remaining_json = remainder;
-      break;
-    }
-    case AZ_JSON_TOKEN_STRING:
-    {
-      json_builder->_internal.need_comma = true;
-      return az_json_builder_append_str(json_builder, token._internal.string);
-    }
-    case AZ_JSON_TOKEN_OBJECT:
-    {
-      AZ_RETURN_IF_NOT_ENOUGH_SIZE(remaining_json, az_span_size(token._internal.span));
-      json_builder->_internal.need_comma = true;
-      remaining_json = az_span_copy(remaining_json, token._internal.span);
-      json_builder->_internal.length += az_span_size(token._internal.span);
-      break;
-    }
-    case AZ_JSON_TOKEN_OBJECT_START:
-    {
-      json_builder->_internal.need_comma = false;
-      remaining_json = az_span_copy_u8(remaining_json, '{');
-      json_builder->_internal.length++;
-      break;
-    }
-    case AZ_JSON_TOKEN_OBJECT_END:
-    {
-      json_builder->_internal.need_comma = true;
-      remaining_json = az_span_copy_u8(remaining_json, '}');
-      json_builder->_internal.length++;
-      break;
-    }
-    case AZ_JSON_TOKEN_ARRAY_START:
-    {
-      json_builder->_internal.need_comma = false;
-      remaining_json = az_span_copy_u8(remaining_json, '[');
-      json_builder->_internal.length++;
-      break;
-    }
-    case AZ_JSON_TOKEN_ARRAY_END:
-    {
-      json_builder->_internal.need_comma = true;
-      remaining_json = az_span_copy_u8(remaining_json, ']');
-      json_builder->_internal.length++;
-      break;
-    }
-    case AZ_JSON_TOKEN_SPAN:
-    {
-      json_builder->_internal.need_comma = true;
-      return _az_json_builder_write_span(json_builder, token._internal.span);
-    }
-    default:
-    {
-      return AZ_ERROR_ARG;
-    }
+    required_size++; // For the leading comma separator.
+  }
+
+  int32_t index_of_first_escaped_char = -1;
+  required_size += _az_json_builder_escaped_length(name, &index_of_first_escaped_char);
+
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(remaining_json, required_size);
+
+  if (json_builder->_internal.need_comma)
+  {
+    remaining_json = az_span_copy_u8(remaining_json, ',');
   }
```

Comment:
I prefer to keep it inline right after the required_size calculation, especially since it is just two lines. This is  similar pattern to what folks do in IoT.